tests/benchmark/bm_batch_iterator.cpp

```diff
@@ -83,33 +96,125 @@ BENCHMARK_DEFINE_F(BM_BatchIterator, get_10000_results_BF)(benchmark::State &st)
+                    correct++;
+                    break;
+                }
+            }
+            VecSimQueryResult_IteratorFree(bf_it);
+        }
+        VecSimQueryResult_IteratorFree(hnsw_it);
+        VecSimQueryResult_Free(hnsw_results);
+    }
+    st.counters["Recall"] = (float)correct / total_res_num;
+    VecSimQueryResult_Free(bf_results);
+    VecSimBatchIterator_Free(batchIterator);
+
     for (auto _ : st) {
-        VecSimBatchIterator *batchIterator = VecSimBatchIterator_New(hnsw_index, query.data());
-        size_t res_num = 0;
+        batchIterator = VecSimBatchIterator_New(hnsw_index, query.data());
+        res_num = 0;
         while (VecSimBatchIterator_HasNext(batchIterator)) {
             VecSimQueryResult_List res = VecSimBatchIterator_Next(batchIterator, n_res, BY_SCORE);
             res_num += VecSimQueryResult_Len(res);
-            if (res_num == 10000) {
+            if (res_num == total_res_num) {
                 break;
             }
         }
         VecSimBatchIterator_Free(batchIterator);
     }
 }
 
-BENCHMARK_REGISTER_F(BM_BatchIterator, get_10000_results_HNSW)
-    ->Arg(100)
+BENCHMARK_REGISTER_F(BM_BatchIterator, BatchedSearch_HNSW)
+    // {batch_size, total_results_count}
+    ->Args({10, 1000})
+    ->Args({100, 1000})
+    ->Args({100, 10000})
+    ->Args({1000, 10000})
+    ->Unit(benchmark::kMillisecond);
+
+BENCHMARK_DEFINE_F(BM_BatchIterator, TopK_BF)(benchmark::State &st) {
+    size_t k = st.range(0);
+    for (auto _ : st) {
+        VecSimIndex_TopKQuery(bf_index, query.data(), k, nullptr, BY_SCORE);
+    }
+}
+
+BENCHMARK_DEFINE_F(BM_BatchIterator, TopK_HNSW)(benchmark::State &st) {
+    size_t k = st.range(0);
+    auto bf_results = VecSimIndex_TopKQuery(bf_index, query.data(), k, nullptr, BY_SCORE);
+    auto hnsw_results = VecSimIndex_TopKQuery(hnsw_index, query.data(), k, nullptr, BY_SCORE);
+
+    // Measure recall
+    auto hnsw_it = VecSimQueryResult_List_GetIterator(hnsw_results);
+    size_t correct = 0;
+    while (VecSimQueryResult_IteratorHasNext(hnsw_it)) {
+        auto hnsw_res_item = VecSimQueryResult_IteratorNext(hnsw_it);
+        auto bf_it = VecSimQueryResult_List_GetIterator(bf_results);
+        while (VecSimQueryResult_IteratorHasNext(bf_it)) {
+            auto bf_res_item = VecSimQueryResult_IteratorNext(bf_it);
+            if (VecSimQueryResult_GetId(hnsw_res_item) == VecSimQueryResult_GetId(bf_res_item)) {
+                correct++;
+                break;
+            }
+        }
+        VecSimQueryResult_IteratorFree(bf_it);
+    }
+    VecSimQueryResult_IteratorFree(hnsw_it);
+    st.counters["Recall"] = (float)correct / k;
+
+    VecSimQueryResult_Free(bf_results);
+    VecSimQueryResult_Free(hnsw_results);
+    for (auto _ : st) {
+        VecSimIndex_TopKQuery(hnsw_index, query.data(), k, nullptr, BY_SCORE);
+    }
+}
+
+// Register the function as a benchmark
+BENCHMARK_REGISTER_F(BM_BatchIterator, TopK_BF)
```

Comment:
what is the purpose of such high numbers?

Reply:
To compare the batched search to a regular top k search (i.e.,  what is the penalty for using batches instead of asking for all the results in one take?)